test/conftest.py

```diff
@@ -99,21 +99,30 @@ def cluster_con(hosts: List[str], port: int):
 
 @pytest.mark.asyncio
 @pytest.fixture(scope="session")
-async def harness(event_loop, request):
+async def harness_internal(event_loop, request):
     """Session fixture to set up client object for communicating with the Cluster API.
        Pass the Unix socket path where the Harness server API is listening.
        Pass a function to create driver connections.
        Test cases (functions) should not use this fixture.
     """
     harness_int = HarnessCli(request.config.getoption('harness_api'), cluster_con)
     await harness_int.start()
-    await harness_int.driver_connect()
     yield harness_int
     harness_int.driver_close()   # Close after last test case
 
+@pytest.mark.asyncio
+@pytest.fixture(scope="function")
+async def harness(request, harness_internal):
+    """Per test fixture to notify Harness client object when tests begin so it can
+    perform checks for cluster state.
+    """
+    await harness_internal.before_test(request.node.name)
+    yield harness_internal
+    await harness_internal.after_test(request.node.name)
+
 # "cql" fixture: set up client object for communicating with the CQL API.
 # We use scope="session" so that all tests will reuse the same client object.
-@pytest.fixture(scope="session")
+@pytest.fixture(scope="function")
```

Comment:
This (making a fixture function instead of session) is a big change, which makes some sense for other reasons (see https://github.com/scylladb/scylla/issues/9599) but I don't understand the reason here and if we don't do too much work on each and every test.
But as long as this doesn't touch cql-pytest and slow down all its quick tests, I won't fight it.

Reply:
For topology tests, clusters are used per test _case_. So cql connection needs to be updated. I thought it's overkill to add logic to keep the driver connected when almost all cases will use a fresh cluster.

BTW in the refactor I forgot to update the comment to reflect this.

Reply:
Now this lives in `test/topology/conftest.py` and nothing else is affected. All topology tests need the driver connection to be managed per test case.